Stage_2

AprilYUZhang opened the pull request.

gregorgorjanc commented on docs/source/usage.rst:
```diff
@@ -300,9 +302,9 @@ Example:
 =====================
 
-|Software| outputs three model parameter files, ``.maf``, ``.seqError``, ``.genoError``. These give the minor allele frequency, sequencing error rates, and genotyping error rates used. All three files contain a single column with an entry for each marker.
+|Software| outputs three model parameter files, ``.alt_allele_prob.txt``, ``.seqError``, ``.genoError``. These give the minor allele frequency, sequencing error rates, and genotyping error rates used. All three files contain a single column with an entry for each marker.
```
@AprilYUZhang we should change docs for these error file outputs too, right?

AprilYUZhang replied:
Yes, it will be next step. I changed the order a little.

gregorgorjanc replied:
@AprilYUZhang is now the time to address these file name changes (`seqErrors` and `genoErrors`)?

gregorgorjanc commented on docs/source/usage.rst:
```diff
@@ -108,16 +110,18 @@ Peeling arguments
+The ``-geno_error_prob``, ``-seq_error_prob`` and ``-length`` arguments control some of the model parameters used in the model. ``-seq_error_prob`` must not be zero. |Software| is robust to deviations in genotyping error rate and sequencing error rate so it is not recommended to use these options unless large deviations from the default are known. Changing the ``-length`` argument to match the ge ...
 
-The ``-esterrors`` option estimated the genotyping error rate based on observed information, this option is generally not necessary and can increase runtime. ``-estmaf`` estimates the minor allele frequency after each peeling cycle. This option can be useful if there are a large number of non-genotyped founders. 
+The ``-est_geno_error_prob`` and ``-est_seq_error_prob`` option estimated the genotyping error rate and the sequencing error rate separately based on observed information, this option is generally not necessary and can increase runtime. ``-est_alt_allele_prob`` estimates the alternative allele probability after each peeling cycle. This option can be useful if there are a large number of non-genot ...
```
@AprilYUZhang change to

```
The ``-est_geno_error_prob`` and ``-est_seq_error_prob`` options estimate the genotyping error rate and the sequencing error rate based on miss-match between observed and inferred states. This option is generally not necessary and can increase runtime. ``-est_alt_allele_prob`` estimates the alternative allele probability after each peeling cycle. This option can be useful if there are a large number of non-genotyped founders.
```

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -357,22 +385,28 @@ def getArgs():
         action="store_true",
         required=False,
-        help="Flag to re-estimate the minor allele frequency after each peeling cycle.",
+        help="Flag to re-estimate the alternative allele probability after each peeling cycle.",
```
@AprilYUZhang change to `probabilities` since we do it for multiple loci. Check also elsewhere.

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -28,9 +28,11 @@ def runPeelingCycles(pedigree, peelingInfo, args, singleLocusMode=False):
         # PeelingUpdates.updateSeg(peelingInfo) #Option currently disabled.
 
-        if args.esterrors:
+        if args.est_geno_error_prob:
```
@AprilYUZhang should we do an "or" statement here? `if args.est_geno... or args.est_seq...`?

AprilYUZhang replied:
I agree, and `if .. or ..:` is better.

gregorgorjanc reviewed (changes requested):
@AprilYUZhang I requested minor tweaks

AprilYUZhang commented:
I found a recombination rate table including human, pig, rat, and so on. Aimed at different species, we need different recombination rates, It may be useful  [Dumont BL, Payseur BA. Evolution of the genomic rate of recombination in mammals. Evolution. 62:276, 2008.](https://onlinelibrary.wiley.com/doi/10.1111/j.1558-5646.2007.00278.x)

AprilYUZhang commented:
In commit # 0632f47, I copied the `get_input_options()` from tinyhouse, but I just changed the input name when it parsed parameters. The reason why I didn't change the name in the `parse_dictionary`, `pedigree, genotypes, seqfile, bfile` are involved many times in the program calls of tinyhouse.

gregorgorjanc commented on docs/source/usage.rst:
```diff
@@ -108,7 +108,7 @@ Peeling arguments
       -nophasefounders      A flag phase a heterozygous allele in one of the
                             founders (if such an allele can be found).
-      -sexchrom             A flag to that this is a sex chromosome. Sex needs to
+      -sex_chrom            A flag to that this is a sex chromosome. Sex needs to
```
@AprilYUZhang please modify to `A flag to indicate that input data is for a sex chromosome. Sex needs to`

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -488,7 +488,7 @@ def getArgs():
+        "-sex_chrom",
         action="store_true",
         required=False,
         help="A flag to that this is a sex chromosome. Sex needs to be given in the pedigree file. This is currently an experimental option.",
```
@AprilYUZhang use the same help string tweak as above

gregorgorjanc commented on src/tinypeel/Peeling/PeelingIO.py:
```diff
@@ -75,10 +75,10 @@ def writeOutParamaters(peelingInfo):
 
 
-def writeGenotypes(pedigree, genoProbFunc):
+def writeGenotypes(pedigree, genoProbFunc, isSexChrom):
```
@AprilYUZhang I understand what you are doing here, but this is a departure from how you handled the option name changes before. Is there a particular reason that you are now picking a different strategy. Doing that makes code even more non-uniform and if you do this a couple of times, the code style becomes impossible to follow.

gregorgorjanc replied:
@AprilYUZhang OK I propose the following - when you encounter such inconsistencies, open an issue so we can later fix them. By finding these problems, you have already spent quite a bit of time understanding the logic so if you open an issue and explain this in the issue content, then that effort is not lost and it easier to pick up at a later stage when you get back to the issue. OK?

gregorgorjanc reviewed (changes requested):
@AprilYUZhang see a minor and a major comment

gregorgorjanc commented:
> I found a recombination rate table including human, pig, rat, and so on. Aimed at different species, we need different recombination rates, It may be useful [Dumont BL, Payseur BA. Evolution of the genomic rate of recombination in mammals. Evolution. 62:276, 2008.](https://onlinelibrary.wiley.com/doi/10.1111/j.1558-5646.2007.00278.x)

Thanks for sharing. There are now even newer recombination rates for some species.

gregorgorjanc commented:
> In commit # [0632f47](https://github.com/AlphaGenes/AlphaPeel/commit/0632f47a3492b06a9f75491f08136c41c61146ce), I copied the `get_input_options()` from tinyhouse, but I just changed the input name when it parsed parameters. The reason why I didn't change the name in the `parse_dictionary`, `pedigree, genotypes, seqfile, bfile` are involved many times in the program calls of tinyhouse.

@AprilYUZhang @XingerTang I see you are challenged quite a bit by trying to keep `tinyhouse` intact, while changing `AlphaPeel`. I spoke to Hannes today (@bryo-han) and he was of the view that you are effectively doing double work - now on AlphaPeel side and then in the future we will have to fix `tinyhouse` again. What about actually breaking `tinyhouse` for other programs? All other programs bring in `tinyhouse` dependency that is linked to a particular commit via a git submodule. So, if we push development of `tinyhouse` further, we will not immediately break other programs. Once we do the changes we need in AlphaPeel and tinyhouse we can always go an fix any other AlphaX program so that it uses the new `tinyhouse` version. This should simplify things a lot!? Thoughts?

gregorgorjanc commented on src/tinypeel/Peeling/PeelingIO.py:
```diff
@@ -68,6 +68,7 @@ def writeOutParamaters(peelingInfo):
 
     np.savetxt(args.out + ".geno_error_prob.txt", peelingInfo.genoError, fmt="%f")
     np.savetxt(args.out + ".seq_error_prob.txt", peelingInfo.seqError, fmt="%f")
+    np.savetxt(args.out + ".rec_prob.txt", np.empty((1,1)), fmt="%f") # not be realized, just as a placeholder
```
@AprilYUZhang Please open an issue so we don't forget to come back to this code. It's easy to forget such tiny TODOs in the code. Also, we are likely storing recombination length of the chromosome and use it to calculate recombination rate between loci - it is that rate that should be outputted here. Make a note of this in the issue.

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -238,6 +238,36 @@ def get_probability_options():
+    parse_dictionary["phasefile"] = lambda parser: parser.add_argument('-phasefile', default=None, required=False,
+                                                                       type=str, nargs="*",
+                                                                       help='A phase file in AlphaGenes format.')
+    parse_dictionary["startsnp"] = lambda parser: parser.add_argument('-startsnp', default=None, required=False,
```
@AprilYUZhang in https://github.com/AlphaGenes/AlphaPeel/issues/108 we discussed to use `-start_snp` and `-stop_snp`

AprilYUZhang replied:
Yes, these options are in the later order of the changing list.

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -238,6 +238,36 @@ def get_probability_options():
+    parse_dictionary["pedigree"] = lambda parser: parser.add_argument('-ped_file', default=None, required=False,
+                                                                      type=str, nargs="*",
+                                                                      help='A pedigree file in AlphaGenes format.')
+    parse_dictionary["phasefile"] = lambda parser: parser.add_argument('-phasefile', default=None, required=False,
```
@AprilYUZhang I think we have missed this in our meeting - in line with the https://github.com/AlphaGenes/AlphaPeel/issues/89 we should call this `-hap_file`

AprilYUZhang replied:
Ok, I will add it.

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -238,6 +238,36 @@ def get_probability_options():
+                                                                      help='A pedigree file in AlphaGenes format.')
+    parse_dictionary["phasefile"] = lambda parser: parser.add_argument('-phasefile', default=None, required=False,
+                                                                       type=str, nargs="*",
+                                                                       help='A phase file in AlphaGenes format.')
```
`A haplotype file in AlphaGenes format.`

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -238,6 +238,36 @@ def get_probability_options():
+    parse_dictionary["startsnp"] = lambda parser: parser.add_argument('-startsnp', default=None, required=False,
+                                                                      type=int,
+                                                                      help="The first marker to consider. The first marker in the file is marker '1'. Default: 1.")
+    parse_dictionary["stopsnp"] = lambda parser: parser.add_argument('-stopsnp', default=None, required=False, type=int,
```
@AprilYUZhang as with `-start_snp`

gregorgorjanc commented on src/tinypeel/Peeling/PeelingIO.py:
```diff
@@ -68,7 +68,9 @@ def writeOutParamaters(peelingInfo):
     np.savetxt(args.out + ".geno_error_prob.txt", peelingInfo.genoError, fmt="%f")
     np.savetxt(args.out + ".seq_error_prob.txt", peelingInfo.seqError, fmt="%f")
-    np.savetxt(args.out + ".rec_prob.txt", np.empty((1,1)), fmt="%f") # not be realized, just as a placeholder
+    np.savetxt(
```
@AprilYUZhang as mentioned in another comment - we should flag this line in an issue so we get back to it at some point

XingerTang commented:
> > In commit # [0632f47](https://github.com/AlphaGenes/AlphaPeel/commit/0632f47a3492b06a9f75491f08136c41c61146ce), I copied the `get_input_options()` from tinyhouse, but I just changed the input name when it parsed parameters. The reason why I didn't change the name in the `parse_dictionary`, `pedigree, genotypes, seqfile, bfile` are involved many times in the program calls of tinyhouse.
> 
> @AprilYUZhang @XingerTang I see you are challenged quite a bit by trying to keep `tinyhouse` intact, while changing `AlphaPeel`. I spoke to Hannes today (@bryo-han) and he was of the view that you are effectively doing double work - now on AlphaPeel side and then in the future we will have to fix `tinyhouse` again. What about actually breaking `tinyhouse` for other programs? All other programs bring in `tinyhouse` dependency that is linked to a particular commit via a git submodule. So, if we push development of `tinyhouse` further, we will not immediately break other programs. Once we do the changes we need in AlphaPeel and tinyhouse we can always go an fix any other AlphaX program so that it uses the new `tinyhouse` version. This should simplify things a lot!? Thoughts?

@gregorgorjanc 
I understand that we may need to do these changes in `tinyhouse` later as well, but @AprilYUZhang has already finished all the modifications related to the `tinyhouse`. And the update of the option names of `tinyhouse` is more difficult than that of `alphapeel` as the option names are used everywhere among the 10 files of `tinyhouse`. It would be more likely for us to miss some of them while modifying the code. 

Meanwhile, `tinyhouse` may still have some bugs unfixed, using different versions of it means we need to introduce changes to different versions of it every time we want to make an update. 

I think it would be better if we just stick with this "double work"  for a while and then when we are ready to make some big changes to the `tinyhouse`, we can do everything at once.

gregorgorjanc commented:
> the update of the option names of `tinyhouse` is more difficult than that of `alphapeel` as the option names are used everywhere among the 10 files of `tinyhouse`. It would be more likely for us to miss some of them while modifying the code.

@XingerTang why would this be hard? It's mostly an input and maybe some variable name change, which is less than doing extra logic, particularly in https://github.com/AlphaGenes/AlphaPeel/pull/122/commits/5821d432e533277590a2b620dfbb4d8feb76708e. The part that I am worried is that while these current changes are not large, they are adding extra logic to `AlphaPeel` and when we decide to polish `tinyhouse` we will forget that some of this extra logic should be removed from `AlphaPeel`.

XingerTang commented:
> > the update of the option names of `tinyhouse` is more difficult than that of `alphapeel` as the option names are used everywhere among the 10 files of `tinyhouse`. It would be more likely for us to miss some of them while modifying the code.
> 
> @XingerTang why would this be hard? It's mostly an input and maybe some variable name change, which is less than doing extra logic, particularly in [5821d43](https://github.com/AlphaGenes/AlphaPeel/commit/5821d432e533277590a2b620dfbb4d8feb76708e). The part that I am worried is that while these current changes are not large, they are adding extra logic to `AlphaPeel` and when we decide to polish `tinyhouse` we will forget that some of this extra logic should be removed from `AlphaPeel`.

It would be hard doesn't because it's hard to write anything, but the code of `AlphaPeel` and `tinyhouse` are messy, and it is hard to track where the option names go. For those option names that are adding/modifying functionality, it is not enough to do "search and replace" as we need to find everywhere it is used and understand how it is used.
In the above, @AprilYUZhang found where the errors are updated and modified the line of the condition for estimating genotype error and sequence error before that line, 
from 
```
        if args.esterrors:
            PeelingUpdates.updatePenetrance(pedigree, peelingInfo)
```
to
```
        if args.est_geno_error_prob or args.est_seq_error_prob:
            PeelingUpdates.updatePenetrance(pedigree, peelingInfo)
```
but that is not enough, we need to look up quite closely to see which line corresponds to the update of genotype error and which line corresponds to the update of sequence error inside ``PeelingUpdates.updatePenetrance(pedigree, peelingInfo)`` and add additional conditions for them, even though there is no ``args.esterrors`` in it. We need to be quite familiar with the code to do all the changes without missing anything. And, when we are facing `tinyhouse`, there are 10 files. It is easy to lose track of where and how some values are passed from one place to another, but we need to understand the uses of all of them before making changes. That is the time-consuming part. 
Also, we don't want code inside `tinyhouse` inconsistent, so we cannot modify only the `AlphaPeel` related part of `tinyhouse`. 

When we polish `tinyhouse`, we also have to check every place of `AlphaPeel` that is related to `tinyhouse` to avoid inconsistency, so we won't miss any extra logic.

gregorgorjanc commented:
@XingerTang @AprilYUZhang I think I understand now. Ok, let's continue and be mindful of the steps beyond the current step. I appreciate your efforts!

AprilYUZhang commented on src/tinypeel/Peeling/PeelingUpdates.py:
`if args.est_geno_error_prob or args.est_seq_error_prob:` doesn't work, right? I'm not sure if you mean  problem happens to  `def runPeelingCycles(pedigree, peelingInfo, args, singleLocusMode=False):
 `.

AprilYUZhang commented:
> > > the update of the option names of `tinyhouse` is more difficult than that of `alphapeel` as the option names are used everywhere among the 10 files of `tinyhouse`. It would be more likely for us to miss some of them while modifying the code.
> > 
> > 
> > @XingerTang why would this be hard? It's mostly an input and maybe some variable name change, which is less than doing extra logic, particularly in [5821d43](https://github.com/AlphaGenes/AlphaPeel/commit/5821d432e533277590a2b620dfbb4d8feb76708e). The part that I am worried is that while these current changes are not large, they are adding extra logic to `AlphaPeel` and when we decide to polish `tinyhouse` we will forget that some of this extra logic should be removed from `AlphaPeel`.
> 
> @gregorgorjanc It would be hard doesn't because it's hard to write anything, but the code of `AlphaPeel` and `tinyhouse` are messy, and it is hard to track where the option names go. For those option names that are adding/modifying functionality, it is not enough to do "search and replace" as we need to find everywhere it is used and understand how it is used. In the above, @AprilYUZhang found where the errors are updated and modified the line of the condition for estimating genotype error and sequence error before that line, from
> 
> ```
>         if args.esterrors:
>             PeelingUpdates.updatePenetrance(pedigree, peelingInfo)
> ```
> 
> to
> 
> ```
>         if args.est_geno_error_prob or args.est_seq_error_prob:
>             PeelingUpdates.updatePenetrance(pedigree, peelingInfo)
> ```
> 
> but that is not enough, we need to look up quite closely to see which line corresponds to the update of genotype error and which line corresponds to the update of sequence error inside `PeelingUpdates.updatePenetrance(pedigree, peelingInfo)` and add additional conditions for them, even though there is no `args.esterrors` in it. We need to be quite familiar with the code to do all the changes without missing anything. And, when we are facing `tinyhouse`, there are 10 files. It is easy to lose track of where and how some values are passed from one place to another, but we need to understand the uses of all of them before making changes. That is the time-consuming part. Also, we don't want code inside `tinyhouse` inconsistent, so we cannot modify only the `AlphaPeel` related part of `tinyhouse`.
> 
> When we polish `tinyhouse`, we also have to check every place of `AlphaPeel` that is related to `tinyhouse` to avoid inconsistency, so we won't miss any extra logic.

Thank you! you explain so clearly.

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -213,6 +213,107 @@ def generateSingleLocusSegregation(peelingInfo, pedigree, args):
+        required=False,
+        type=str,
+        nargs="*",
+        help="A file in plink (binary) format. Only stable on Linux).",
```
@AprilYUZhang remove last `)`

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -213,6 +213,107 @@ def generateSingleLocusSegregation(peelingInfo, pedigree, args):
+        help="A file in AlphaGenes format.",
+    )
+    parse_dictionary["reference"] = lambda parser: parser.add_argument(
+        "-reference",
```
@AprilYUZhang do we mention this file at all in documentation? I think we are not. Please open an issue to sort this.

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -213,6 +213,107 @@ def generateSingleLocusSegregation(peelingInfo, pedigree, args):
+        help="A pedigree file in AlphaGenes format.",
+    )
+    parse_dictionary["phasefile"] = lambda parser: parser.add_argument(
+        "-phasefile",
```
@AprilYUZhang we will call this file diffrently!?

AprilYUZhang replied:
Yeah, but I will open a new commit to fix it.

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -213,6 +213,107 @@ def generateSingleLocusSegregation(peelingInfo, pedigree, args):
+        help="A phase file in AlphaGenes format.",
+    )
+    parse_dictionary["startsnp"] = lambda parser: parser.add_argument(
+        "-startsnp",
```
@AprilYUZhang should not this be `-start_snp`?

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -213,6 +213,107 @@ def generateSingleLocusSegregation(peelingInfo, pedigree, args):
+        help="The first marker to consider. The first marker in the file is marker '1'. Default: 1.",
+    )
+    parse_dictionary["stopsnp"] = lambda parser: parser.add_argument(
+        "-stopsnp",
```
@AprilYUZhang should not this be `-stop_snp`?

gregorgorjanc reviewed (changes requested):
@AprilYUZhang I reviewed your recent changes and requested some modifications.

AprilYUZhang commented:
I find ".called." is not fixed in `writeBinaryCalledGenotypes`. Although it hasn’t been realized, do we need to change it? @gregorgorjanc @XingerTang

XingerTang commented:
> I find ".called." is not fixed in `writeBinaryCalledGenotypes`. Although it hasn’t been realized, do we need to change it? @gregorgorjanc @XingerTang

@AprilYUZhang It's not important for now. I prefer to leave that to the future.

gregorgorjanc commented:
> > I find ".called." is not fixed in `writeBinaryCalledGenotypes`. Although it hasn’t been realized, do we need to change it? @gregorgorjanc @XingerTang
> 
> @AprilYUZhang It's not important for now. I prefer to leave that to the future.

@AprilYUZhang open an issue and we get back to it in future

AprilYUZhang commented:
I updated the summary of changes #113

gregorgorjanc commented on src/tinypeel/Peeling/PeelingIO.py:
```diff
@@ -66,47 +66,59 @@ def readInSeg(pedigree, fileName, start=None, stop=None):
+    np.savetxt(
+        args.out_file + ".rec_prob.txt", np.empty((1, 1)), fmt="%f"
+    )  # not be realized, just as a placeholder
+    # np.savetxt(args.out_file + ".trans", peelingInfo.transmissionRate, fmt = "%f")
```
@AprilYUZhang do you know what exactly is in `peelingInfo.transmissionRate`? I wonder if this anyway connected to mutations? Open an issue and we can look at it later.

AprilYUZhang replied:
```
def setupTransmission(length, peelingInfo):
    if peelingInfo.positions is None:
        localMap = np.linspace(0, 1, num=peelingInfo.nLoci, dtype=np.float32)
    else:
        localMap = (
            peelingInfo.positions / peelingInfo.positions[-1]
        )  # This should be sorted. Need to add in code to check.
    for i in range(peelingInfo.nLoci - 1):
        distance = localMap[i + 1] - localMap[i]
        distance = distance * length
        peelingInfo.transmissionRate[i] = distance
```
I found the code that calculates the distance between a locus and an adjacent locus. It seems to be used in the updateSeg(), but now it has been abandoned.
```
# def updateSeg(peelingInfo):
#     # Idea: Split the chromosome into ~10 blocks with slightly different starts/stops. Estimate at each one and then sum.
#     # Not sure if this is the best way, but probably will work okay.

#     distances = [
#         estDistanceFromBreaks(0, peelingInfo.nLoci - 1, nBreaks, peelingInfo)
#         for nBreaks in [5, 10, 20]
#     ]
#     distance = np.mean(distances[0:2])
#     print(distances, ":", distance)

#     setupTransmission(distance, peelingInfo)
```

gregorgorjanc replied:
@AprilYUZhang the above help flag has still `-outputkey` mentioned, while it should be `out_id_order`

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -213,19 +213,166 @@ def generateSingleLocusSegregation(peelingInfo, pedigree, args):
+        help="Sequencing error rate. Default: 0.001.",
+    )
+    parse_dictionary["recombination"] = lambda parser: parser.add_argument(
+        "-recomb",
```
@AprilYUZhang should this be `rec_length`?

gregorgorjanc replied:
@AprilYUZhang hmm, the help text is the same as for `rec_length` as is the default suggesting we are confusing something here. At any rate, looking at the table https://github.com/AlphaGenes/AlphaPeel/issues/113 we have `rec_length` as input for the genetic length of the chromosome, `est_rec_prob` to flag estimation of the recombination rate between loci, and `.rec_prob` for file output (I made this clear in https://github.com/AlphaGenes/AlphaPeel/issues/113). There is no `recombination` option that I am aware of, but let me know if I am confusing something.

https://github.com/AlphaGenes/AlphaPeel/issues/97
https://github.com/AlphaGenes/AlphaPeel/issues/104

AprilYUZhang replied:
I get your means; I made the mistake that rec_prob is not an option rather than an output file name.

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -213,19 +213,166 @@ def generateSingleLocusSegregation(peelingInfo, pedigree, args):
+        default="id",
+        required=False,
+        type=str,
+        help='Determines the order in which individuals are ordered in the output file based on their order in the corresponding input file. Animals not in the input file are placed at the end of the file and sorted in alphanumeric order. These animals can be surpressed with the "-onlykeyed" option. Options: id, pedigree, genotypes, sequence, segregation. Defualt: id.',
```
Change `-onlykeyed` to `-out_id_only`

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -213,19 +213,166 @@ def generateSingleLocusSegregation(peelingInfo, pedigree, args):
+        "-out_id_only",
+        action="store_true",
+        required=False,
+        help='Flag to surpress the animals who are not present in the file used with -outputkey. Also surpresses "dummy" animals.',
```
Change `outputkey` to `???`

AprilYUZhang replied:
I'm not sure if you pointed at `writekey`, it has been changed to `out_id_order`.

AprilYUZhang replied:
It should be "-out_id_order". In the documents, it is right.

gregorgorjanc commented on src/tinypeel/tinypeel.py:
```diff
@@ -213,19 +213,166 @@ def generateSingleLocusSegregation(peelingInfo, pedigree, args):
+def get_multithread_options():
+    parse_dictionary = dict()
+    parse_dictionary["iothreads"] = lambda parser: parser.add_argument(
+        "-n_io_threads",
```
@AprilYUZhang please ensure that documentation files are updated too - I don't think I have spotted this option being documented and I did not pay attention before either.

gregorgorjanc reviewed (changes requested):
@AprilYUZhang minor changes/revisions requested

AprilYUZhang commented:
the problem of these two options are similar toIn  commit #[0632f47](https://github.com/AlphaGenes/AlphaPeel/commit/0632f47a3492b06a9f75491f08136c41c61146ce). I just changed the part parsing argument rather than all. When we consider to change tinyhouse, it will be fixed in the whole software. 

> In commit # [0632f47](https://github.com/AlphaGenes/AlphaPeel/commit/0632f47a3492b06a9f75491f08136c41c61146ce), I copied the `get_input_options()` from tinyhouse, but I just changed the input name when it parsed parameters. The reason why I didn't change the name in the `parse_dictionary`, `pedigree, genotypes, seqfile, bfile` are involved many times in the program calls of tinyhouse.

gregorgorjanc commented:
> the problem of these two options are similar toIn commit #[0632f47](https://github.com/AlphaGenes/AlphaPeel/commit/0632f47a3492b06a9f75491f08136c41c61146ce). I just changed the part parsing argument rather than all. When we consider to change tinyhouse, it will be fixed in the whole software.
> 
> > In commit # [0632f47](https://github.com/AlphaGenes/AlphaPeel/commit/0632f47a3492b06a9f75491f08136c41c61146ce), I copied the `get_input_options()` from tinyhouse, but I just changed the input name when it parsed parameters. The reason why I didn't change the name in the `parse_dictionary`, `pedigree, genotypes, seqfile, bfile` are involved many times in the program calls of tinyhouse.

@AprilYUZhang I don't know to which question/comment you are referring to with this answer. Can you clarity? Is this related to my comment in this thread https://github.com/AlphaGenes/AlphaPeel/pull/122#discussion_r1399369654?

gregorgorjanc commented:
@AprilYUZhang I think this looks very good all in all! Are there any more planned changes - it seems you have done all of the points in https://github.com/AlphaGenes/AlphaPeel/issues/113!? 

@XingerTang I would appreciate quick code review from your end before we merge this in.

XingerTang commented on src/tinypeel/tinypeel.py:
```diff
@@ -213,19 +213,160 @@ def generateSingleLocusSegregation(peelingInfo, pedigree, args):
+        help="A random seed to use for debugging.",
+    )
+
+    return parse_dictionary
```
@AprilYUZhang 
Even though you keep the keys of the `parse_dictionary` unchanged, you still need to change how you access the values via `args` inside the main program. 

I did a test program, it basically does the same thing as those functions in `tinyhouse.InputOutput`

```python
import argparse
import sys

parser = argparse.ArgumentParser(description = 'Test')

def add_arguments_from_dictionary(parser, arg_dict, options = None):
    if options is None:
        for key, value in arg_dict.items():
            value(parser)
    else:
        for option in options:
            if option in arg_dict:
                arg_dict[option](parser)
            else:
                print("Option not found:", option, arg_dict)

def get_input_options():

    parse_dictionary = dict()
    parse_dictionary["bfile"] = lambda parser: parser.add_argument(
        '-plink_file',   
        default=None, 
        required=False, 
        type=str, 
        nargs="*", 
        help='A file in plink (binary) format. Only stable on Linux).'
        )
    
    return parse_dictionary

input_parser = parser.add_argument_group("Input Options")
add_arguments_from_dictionary(
        input_parser,
        get_input_options(),
        options=[
            "bfile"
        ],
)

args = sys.argv[1:]
args = parser.parse_args(args)
print(f"{args.plink_file=}")
print(f"{args.bfile=}")
```
And I get 
```bash
(base) evie@Xingers-MacBook-Air test % python /Users/evie/test/src/test/testing.py -plink_file "FILE NAME"
args.plink_file=['FILE NAME']
Traceback (most recent call last):
  File "/Users/evie/test/src/test/testing.py", line 43, in <module>
    print(f"{args.bfile=}")
AttributeError: 'Namespace' object has no attribute 'bfile'
```
So the `args` can only know `plink_file` as the name in the command line `plink_file` but not the dictionary key `bfile`.

You still need to change the places where the option names are changed in the main program. I think one way to do this is by redefining the attributes of the `args` 
```python
args = sys.argv[1:]
args = parser.parse_args(args)
args.bfile = args.plink_file
print(f"{args.bfile=}")
```
And with this change I can get
```bash
(base) evie@Xingers-MacBook-Air test % python /Users/evie/test/src/test/testing.py -plink_file "FILE NAME"
args.bfile=['FILE NAME']
```

AprilYUZhang replied:
Maybe today we can talk about it, but why you need use args.bfile in the test? I'm a little confused. Because I don't know what is wrong in the main programs.

XingerTang replied:
@AprilYUZhang We don't need `args.bfile` in the test for now, I just use it as an example to explain why I think you need to add a few more lines to make the code work.

XingerTang commented on src/tinypeel/Peeling/PeelingUpdates.py:
@AprilYUZhang Please add conditions of `if args.est_geno_error_prob:` and `if args.est_seq_error_prob:` before all the genotype error updates and the sequence error updates respectively

AprilYUZhang replied:
I get it

XingerTang commented:
@AprilYUZhang @gregorgorjanc There are still some changes that need to be made before we merge. Please see the above comments.

XingerTang commented:
@AprilYUZhang @gregorgorjanc I just noticed that we still have an `s` in `n_cycles` which is the updated name of `ncycles`. Should we keep it or change it to `n_cycle`?

gregorgorjanc commented:
> @AprilYUZhang @gregorgorjanc I just noticed that we still have an `s` in `n_cycles` which is the updated name of `ncycles`. Should we keep it or change it to `n_cycle`?

@XingerTang well spotted! @AprilYUZhang please check all the options and let us know what you suggest as a unified approach regarding the plural and the associated “s”.

AprilYUZhang commented:
Solved the above problems.

XingerTang commented:
@gregorgorjanc 
Could you rerun this following workflow
<img width="1410" alt="Screen Shot 2023-11-24 at 8 51 05 PM" src="https://github.com/AlphaGenes/AlphaPeel/assets/50447050/ae9accb1-5aae-4c59-9e8e-7a3115c55c89">
By clicking the job bar on the left
<img width="534" alt="Screen Shot 2023-11-24 at 8 51 29 PM" src="https://github.com/AlphaGenes/AlphaPeel/assets/50447050/a6740679-f9a1-48ee-9b8c-63a9da772904">

XingerTang commented:
@gregorgorjanc @AprilYUZhang 
Sorry I forgot the commit here is not up to date and sent the wrong comment, I already deleted that. 
@AprilYUZhang Could you pull the recent changes that just merged into the branch, and let's see how the test go

XingerTang commented on src/tinypeel/Peeling/PeelingIO.py:
```diff
@@ -66,47 +66,59 @@ def readInSeg(pedigree, fileName, start=None, stop=None):
+    # np.savetxt(args.out_file + ".trans", peelingInfo.transmissionRate, fmt = "%f")
+    np.savetxt(
+        args.out_file + ".alt_allele_prob.txt", peelingInfo.alt_allele_prob, fmt="%f"
+    )
```
Please change `peelingInfo.alt_allele_prob` back to `peelingInfo.maf`.

XingerTang commented on src/tinypeel/tinypeel.py:
```diff
@@ -416,28 +562,45 @@ def getArgs():
+    args.onlykeyed = args.out_id_only
+    args.iothreads = args.n_io_thread
+    args.maxthreads = args.n_thread
+
```
Please add a line  of `args.segfile = args.seg_file`.

AprilYUZhang replied:
ok

XingerTang commented on src/tinypeel/tinypeel.py:
```diff
@@ -28,8 +28,7 @@ def runPeelingCycles(pedigree, peelingInfo, args, singleLocusMode=False):
 
-        if args.esterrors:
-            PeelingUpdates.updatePenetrance(pedigree, peelingInfo)
+        PeelingUpdates.updatePenetrance(pedigree, peelingInfo, args)
```
Please add the condition `if args.est_geno_error_prob or args.est_seq_error_prob:` back

XingerTang commented on src/tinypeel/tinypeel.py:
```diff
@@ -416,28 +562,46 @@ def getArgs():
 
 def main():
     args = getArgs()
```
Please also add

```
    if args.start_snp:
            args.start_snp -= 1
    if args.stop_snp:
            args.stop_snp -= 1
```

This is modified from `tinyhouse.InputOutput.parseArgs`, but now we also need it here.

XingerTang commented:
@gregorgorjanc Could you please merge this pull request?

gregorgorjanc commented:
> @gregorgorjanc Could you please merge this pull request?

@XingerTang i lost track of which of the above discussions and requested changes have been addressed l please go over them and resolve each one (maybe the changes have all been achieved?) and then we merge this in,

XingerTang commented:
Yes, all the changes are achieved and every problem is addressed. I think we can merge this in.
